converters/toZigbee.js

```diff
@@ -1429,6 +1429,74 @@ const converters = {
             await entity.read('closuresDoorLock', ['lockState']);
         },
     },
+    pincode_lock: {
+        key: ['pin_code'],
+        convertSet: async (entity, key, value, meta) => {
+            const user = value.user;
+            const pinCode = value.pin_code;
+            if ( isNaN(user) ) {
+                throw new Error('user must be numbers');
+            }
+            if ( !converters.pincode_lock.userInRange(meta, user) ) {
+                throw new Error('user must be in range for device');
+            }
+            if (pinCode === undefined || pinCode === null) {
+                await entity.command(
+                    'closuresDoorLock',
+                    'clearPinCode',
+                    {
+                        'userid': user,
+                    },
+                    getOptions(meta.mapped),
+                );
+            } else {
+                if (isNaN(pinCode)) {
+                    throw new Error('pinCode must be a number or pinCode');
+                }
+                await entity.command(
+                    'closuresDoorLock',
+                    'setPinCode',
+                    {
+                        'userid': user,
+                        'userstatus': 1,
+                        'usertype': 0,
+                        'pincodevalue': pinCode.toString(),
+                    },
+                    getOptions(meta.mapped),
+                );
+            }
+            return {readAfterWriteTime: 200};
+        },
+        convertGet: async (entity, key, meta) => {
+            const user = meta && meta.message && meta.message.pin_code ? meta.message.pin_code.user : undefined;
+            if (user === undefined) {
+                const max = meta.mapped.meta.pinCodeCount;
+                // Get all
+                for (let i = 0; i < max; i++) {
+                    await converters.pincode_lock.getPinCode(entity, i, meta);
+                }
+            } else {
+                if (isNaN(user)) {
+                    throw new Error('user must be numbers');
+                }
+                if (!converters.pincode_lock.userInRange(meta, user)) {
+                    throw new Error('userId must be in range for device');
+                }
+                await converters.pincode_lock.getPinCode(entity, user, meta);
+            }
+        },
+        getPinCode: async (entity, user, meta) => {
```

Comment:
After https://github.com/Koenkk/zigbee-herdsman-converters/pull/1357#discussion_r465982615, this method can also be inlined in `convertGet` (we don't want to expose this).

Reply:
I had some difficulties with this, because with the new assumption that `meta.mapped.meta.pinCodeCount` is always set, we now have to contruct an entire "meta" object in the onEvent context that has this value set, as well as `meta.message.pin_code.user`. I could be wrong, but I do not believe any of the inputs to onEvent have the device meta available.

Reply:
With this helper method, we are able to much more easily call the command from the onEvent handler.

Reply:
Yes I agree, in that case I suggest moving it to `converters/utils.js`.